optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/common/accessor/MemberAccessorFactory.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2018 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.domain.common.accessor;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Field;
+import java.lang.reflect.Member;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+
+import org.optaplanner.core.impl.domain.common.ReflectionHelper;
+
+public class MemberAccessorFactory {
+
+    public static MemberAccessor buildMemberAccessor(Member member, MemberAccessorType memberAccessorType,
+            Class<? extends Annotation> annotationClass) {
+        if (member instanceof Field) {
+            Field field = (Field) member;
+            return new ReflectionFieldMemberAccessor(field);
+        } else if (member instanceof Method) {
+            Method method = (Method) member;
+            MemberAccessor memberAccessor;
+            switch (memberAccessorType) {
```

Comment:
I tried that, but I didn't think it made the logic easier to follow.
Agreed it could be cleaner, but I am not sure how currently.
I' d rather focus on getting rid of the diversification between ReflectionMethodMemberAccessor and ReflectionBeanPropertyMemberAccessor first. (and their lamdba counterparts), but that's a separate issue.